cesql/spec.md

```diff
@@ -56,7 +56,10 @@ producer, or in an intermediary, and it can be implemented using any technology
 The CloudEvents Expression Language assumes the input always includes, but is not limited to, a single valid and
 type-checked CloudEvent instance. An expression MUST NOT mutate the value of the input CloudEvent instance, nor any of
 the other input values. The evaluation of an expression observes the concept of [referential
-transparency][referential-transparency-wiki]. The output of a CESQL expression evaluation is always a _boolean_, an _integer_ or a _string_, and it might include an error.
+transparency][referential-transparency-wiki]. The primary output of a CESQL expression evaluation is always a _boolean_, an _integer_ or a _string_.
+The secondary output of a CESQL expression evaluation is a set of errors which occurred during evaluation. This set MAY be empty, indicating that no
+error occurred during execution of the expression. The value used by CESQL engines to represent an empty set of errors is out of the scope of this
+specification.
```

Comment:
I like this better. Do you think we need to add something like:
> Generation of an error does not halt processing. This specification does not mandate how these errors are exposed to a user.

? Is the first sentence true?

Reply:
I'm not 100% sure here, if you look at section 4.1 it states that:

> When evaluating an expression, the evaluator can operate in two _modes_, in relation to error handling:
> * Fail fast mode: When an error is triggered, the evaluation is interrupted and returns the error, without any result.
> * Complete evaluation mode: When an error is triggered, the evaluation is continued, and the evaluation of the expression returns both the result and the error(s).

But, as far as I can tell the SDKs implement the fail fast mode instead of the complete evaluation mode.

Reply:
w/o any result... is that true even in the SDKs and Knative? I was assuming the was result 'false'.

I'm having a memory of us discussing what happens when a nested expression errors when that expression is part of an AND - doesn't it result in a false for the nested expression and then false again due to that side of the AND being false? Or am I thinking about some other situation?

Reply:
In terms of w/o any result, I think the problem (from a spec standpoint at least) is what type should the return value have? For example, if my query is `LEFT(missingattribute, 10)`, then should I get `(false, missing attribute error)` as my return value? Does it make more sense to return `("", missing attribute error)`?

But, similarly if my expression is `LEFT(missingattribute, 10) LIKE prefix%`, then having a return value of `(false, missing attribute error)` would be what I would expect while a return value of `("", missing attribute error)` would not be what I expect.

Perhaps it makes sense to say something along the lines of "Fail fast mode: When an error is triggered, the evaluation is interrupted and returns the zero value for the return type of the root operation, along with the error". WDYT @duglin ?